gapii/client/capture.go

```diff
@@ -181,41 +186,110 @@ func (p *Process) Capture(ctx context.Context, start task.Signal, stop task.Sign
 	conn := p.conn
 	defer conn.Close()
 
-	var count siSize
-	started := false
+	buf := make([]byte, 64*1024)
+	var bufSize, bufPos int
+	var varintPos uint
+	var chunkSize uint64
+	var readErr error
+	var count, remain siSize
+	var messageType byte = messageInvalid
+	var bufErr []byte
+	started, stopped := false, false
+mainLoop:
```

Comment:
I think we could simplify this if we make the message reader and writer separate goroutines. It would also get rid of the 500ms delay of sending the stop signal (yikes). I also think the reader can be simplified by not having an extra read buffer, but simply be in a `readHeader(); readData()` loop.

Reply:
+1 for the separate reader and writer goroutines.
Bit confused about "extra read buffer". The old implementation uses io.CopyN, which internally allocates a buffer each time. My implementation allocates only one read buffer and reuses it. I think that's worth to make the garbage collector happy. :)
I also think separate readHeader(); readData() loop would not be more simple because the buffer would have to be handed over between them. To avoid that we would have to read much smaller chunks from the stream, which could lead to much more socket synchronization overhead.

Reply:
The start and stop signals are now in blocking goroutines.
I also have simplified, cleaned up and regrouped the read loop into buffer reading, header parsing and data parsing. It would not simplify the code to use separate functions as they share many state variables.

Reply:
>  I think that's worth to make the garbage collector happy.

No, no it is not. The compiler/runtime/allocator/GC are smarter than you think and short lived temporary objects are often preferred. Even if not, this is not a good reason to make the code more complex.

Internally the TCP connection is already buffered. So, since the header is now fixed size, you can simply create a function that reads and returns a header from the socket by reading the 9 bytes. Then you switch on the header type and process things. This would also allow you to use `io.CopyN` again. All the local variables would go away, since they are just necessary to keep track of read and write pointers in your buffer.